File: algorithm/cpp/trapping-rain-water-ii.cpp
// 407. Trapping Rain Water
// Difficulty: Hard

// Given an m x n matrix of positive integers representing the height of each unit cell in a 2D elevation map,
// compute the volume of water it is able to trap after raining.

// Note:
// Both m and n are less than 110. The height of each unit cell is greater than 0 and is less than 20,000.

// Example:

// Given the following 3x6 height map:
// [
//   [1,4,3,1,3,2],
//   [3,2,1,3,2,4],
//   [2,3,3,2,3,1]
//   ]

// Return 4.

// Time:  O(m * n * log(m + n)) ~ O(m * n * log(m * n))
// Space: O(m * n)

class Solution {
public:
  int trapRainWater(vector<vector<int>>& heightMap) {
    int m = heightMap.size(), n = heightMap[0].size();

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> q;
    vector<vector<bool>> visited(m, vector<bool>(n, false));
    
    for (int i = 0; i < m; ++i) {
      for (int j = 0; j < n; ++j) {
        if (i == 0 || i == m - 1 || j == 0 || j == n - 1) {
          q.push({heightMap[i][j], i * n + j});
          visited[i][j] = true;
        }
      }
    }
    int res = 0, level = INT_MIN;
    while (!q.empty()) {
      auto top = q.top(); q.pop();
      int height = top.first, i = top.second / n, j = top.second % n;
      level = max(level, height);
      for (auto dir : vector<vector<int>>{ {0, -1}, {-1, 0}, {0, 1}, {1, 0} }) {
        int x = i + dir.first, y = j + dir.second;
        if (x >= 0 && x < m && y >= 0 && y < n && !visited[x][y]) {
         
          if (heightMap[x][y] < level) {
            res += level - heightMap[x][y];
          }
          visited[x][y] = true;
          q.push({heightMap[x][y], x * n + y});
        }
      }
    }
    return res;
  }
};

File: data-structure-and-algorithm/typescript/leetcode/src/trapping-rain-water-ii.ts
/*
Given an m x n integer matrix heightMap representing the height of each unit cell in a 2D elevation map, return the volume of water it can trap after raining.



Example 1:


Input: heightMap = [[1,4,3,1,3,2],[3,2,1,3,2,4],[2,3,3,2,3,1]]
Output: 4
Explanation: After the rain, water is trapped between the blocks.
We have two small pounds 1 and 3 units trapped.
The total volume of water trapped is 4.
Example 2:


Input: heightMap = [[3,3,3,3,3],[3,2,2,2,3],[3,2,1,2,3],[3,2,2,2,3],[3,3,3,3,3]]
Output: 10


Constraints:

m == heightMap.length
n == heightMap[i].length
1 <= m, n <= 200
0 <= heightMap[i][j] <= 2 * 104
*/

import {Heap} from './data-structure/Heap';

function trapRainWater(heightGrid: number[][]): number {
  const [rows, cols] = [heightGrid.length, heightGrid[0].length];
  const q = new Heap<number[]>(([x0, y0, z0], [x1, y1, z1]) => z0 - z1);
  const visited: boolean[][] = [...Array(rows)].map(() =>
    Array(cols).fill(false)
  );

  for (let i = 0; i < rows; ++i) {
    for (const [x, y] of [
      [i, 0],
      [i, cols - 1],
    ]) {
      q.push([x, y, heightGrid[x][y]]);
      visited[x][y] = true;
    }
  }

  for (let j = 1; j < cols - 1; ++j) {
    for (const [x, y] of [
      [0, j],
      [rows - 1, j],
    ]) {
      q.push([x, y, heightGrid[x][y]]);
      visited[x][y] = true;
    }
  }

  let result = 0;
  while (!q.isEmpty()) {
    const [x, y, h] = q.pop();
    for (const [dx, dy] of [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ]) {
      const [nx, ny] = [x + dx, y + dy];
      if (nx >= 0 && nx < rows && ny >= 0 && ny < cols && !visited[nx][ny]) {
        result += Math.max(0, h - heightGrid[nx][ny]);
        q.push([nx, ny, Math.max(heightGrid[nx][ny], h)]);
        visited[nx][ny] = true;
      }
    }
  }
  return result;
}
